File: src/tests/4 Justify/main.c
#include <stdio.h>
#include <string.h>
#include <sys/time.h>

char S[2047];             // input
unsigned int n;           // length of S
unsigned int W;           // width of the screen
unsigned T;               // time limit (seconds)
struct timeval stopTime;  // time when we should stop
unsigned long iterations; // amount of finished cycles of computing

unsigned int wc = 0;     // words count
unsigned int ws[2048];   // words' positions in S
unsigned int wl[2048];   // words' lengths
unsigned int line[2048]; // number of words optimal in a line
unsigned int res[2048];  // array of minimal free space for each subsequence

void start();
void reinit();
void calc();
void outputText();
void outputLineWithPadding(char *s);

int main (int argc, char ** argv) {
  if (argc != 2) {
    printf("Time limit parameter required:\n ./main.bin <time>\n");
  } else {
    iterations = 0;
    sscanf(argv[1], "%d", &T);
    printf("Calculating unused space F(W, S) for string S and width W many times during %u seconds\n", T);
    printf("Input W: ");

    scanf("%d\n", &W);

    printf("Input S: ");
    fgets(S, 2047, stdin);
    n = (unsigned int)strlen(S);
    S[--n] = 0; // trim new-line
    printf("\n");
    start();
  }
  return 0;
}

void start() {
  struct timeval now;
  gettimeofday(&stopTime, NULL);
  stopTime.tv_sec += T; // NOW + T seconds
  gettimeofday(&now, NULL);
  while (now.tv_sec <  stopTime.tv_sec ||
        (now.tv_sec == stopTime.tv_sec && now.tv_usec < stopTime.tv_usec)) {
    reinit();
    calc();
    iterations++;
    gettimeofday(&now, NULL);
  }
  iterations--; // last iteration started before time was up but finished after time was up
  // required '>' mark for automatic benchmark
  printf("> %lu iterations\n", iterations);
  printf("Resulting F(W, S) = %u\n", res[0]);
  outputText();
}

void reinit() {
  // input parsing done only once, not counted in measurements
  if (wc == 0) {
    // count words
    wc = 1;
    for (unsigned int i = 0; i < n; i++) {
      if (S[i] == ' ') {
        wc++;
      }
    }

    // split by words
    // creating ws and wl of size wc+1
    unsigned int j = 0;
    ws[j] = 0;
    for (unsigned int i = 0; i < n; i++) {
      if (S[i] == ' ') {
        wl[j] = i - ws[j];
	j++;
	ws[j] = i + 1;
      }
    }
    wl[j] = n - ws[j];
    // EOF marker
    wl[wc] = 0;
    ws[wc] = n + 1;

    // res (allready allocated) is of size wc+2
    res[wc + 1] = 0;
    // line (allready allocated) is of size wc+1
  }
  // no cleanup needed
}

void calc() {
  // no content, no unused space
  res[wc] = 0;
  line[wc] = 0;
  for (int i = wc - 1; i >= 0; i--) {
    // start with a single word
    line[i] = 1;
    // space occupied by words
    int sp = wl[i];
    res[i] = W - sp + res[i + line[i]];

    int newln = line[i];
    int newrs = res[i];
    while (sp <= W &&
           i + newln <= wc) {
      if (newrs <= res[i]) {
        line[i] = newln;
	res[i] = newrs;
      }

      sp += 1 + wl[i + newln];
      newln++;
      newrs = W - sp + res[i + newln];
    }
  }
}

void outputText() {
  printf("\n");
  unsigned int pos = 0;
  while (line[pos] != 0) {
    S[ws[pos + line[pos]] - 1] = 0;
    outputLineWithPadding(S + ws[pos]);
    pos += line[pos];
  }
}

void outputLineWithPadding(char *s) {
  printf(" %s", s);
  for (unsigned int i = strlen(s); i < W; i++) {
    printf("_");
  }
  printf("\n");
}

File: src/tests/4 Justify/main.php
<?php
error_reporting(E_ALL);
$S = "";         // input
$n = 0;          // length of S
$W = 0;          // width of the screen
$T = 0;          // time limit (seconds)
$stopTime = 0;   // time when we should stop
$iterations = 0; // amount of finished cycles of computing

$wc = 0;         // words count
$ws = NULL;      // words' positions in S
$wl = NULL;      // words' lengths
$line = NULL;    // number of words optimal in a line
$res = NULL;     // array of minimal free space for each subsequence

if (count($argv) != 2) {
  print("Time limit parameter required:\n php main.php <time>\n");
} else {
  $iterations = 0;
  $T = $argv[1];
  print("Calculating unused space F(W, S) for string S and width W many times during $T seconds\n");
  print("Input W:\n");
  
  fscanf(STDIN, "%d\n", $W);
  print("Input S:\n");

  $S = trim(fgets(STDIN, 2047));
  $n = strlen($S);
  start();
}

function start() {
  global $stopTime, $iterations, $T, $res, $n, $m;
  $stopTime = microtime(true) + $T; // NOW + T seconds
  while ( microtime(true) < $stopTime) {
    reinit();
    calc();
    $iterations++;
  }
  $iterations--; // last iteration started before time was up but finished after time was up
  // required '>' mark for automatic benchmark
  print("> $iterations iterations\n");
  print("Resulting F(W, S) = {$res[0]}\n");
  outputText();
}

function reinit () {
  global $res, $wc, $wl, $ws, $S, $n, $line;
  // input parsing done only once, not counted in measurements
  if ($res === NULL) {
    // count words
    $wc = 1;
    for ($i = 0; $i < $n; $i++) {
      if ($S[$i] === " ") {
        $wc++;
      }
    }

    // split by words
    // creating ws and wl of size wc+1
    $ws = array();
    $wl = array();
    $j = 0;
    $ws[$j] = 0;
    for ($i = 0; $i < $n; $i++) {
      if ($S[$i] === " ") {
        $wl[$j] = $i - $ws[$j];
	$j++;
	$ws[$j] = $i + 1;
      }
    }
    $wl[$j] = $n - $ws[$j];
    // EOF marker
    $wl[$wc] = 0;
    $ws[$wc] = $n + 1;

    // res is of size wc+2
    $res = array_fill(0, $wc + 2, 0);
    // line is of size wc+1
    $line = array_fill(0, $wc + 1, 0);
  }
  // no cleanup needed
}

function calc() {
  global $res, $wc, $wl, $line, $n, $W;
  // no content, no unused space
  $res[$wc] = 0;
  $line[$wc] = 0;

  for ($i = $wc - 1; $i >= 0; $i--) {
    // start with a single word
    $line[$i] = 1;
    // space occupied by words
    $sp = $wl[$i];
    $res[$i] = $W - $sp + $res[$i + $line[$i]];

    $newln = $line[$i];
    $newrs = $res[$i];

    // try to add one more word
    while ($sp <= $W &&
           $i + $newln <= $wc) {
      if ($newrs <= $res[$i]) {
        $line[$i] = $newln;
	$res[$i] = $newrs;
      }

      $sp += 1 + $wl[$i + $newln];
      $newln++;
      $newrs = $W - $sp + $res[$i + $newln];
    }
  }
}

function outputText() {
  global $line, $W, $S, $ws;
  print("\n");
  $pos = 0;
  while ($line[$pos] !== 0) {
    $end = $ws[$pos + $line[$pos]] - 1;
    $len = $end - $ws[$pos];
    outputLineWithPadding(substr($S, $ws[$pos], $len));
    $pos += $line[$pos];
  }
}

function outputLineWithPadding($s) {
  global $W;
  print(" $s");
  for ($i = strlen($s); $i < $W; $i++) {
    print("_");
  }
  print("\n");
}
